src/python/pants/backend/python/rules/pex.py

```diff
@@ -15,18 +15,65 @@
   DirectoriesToMerge,
   DirectoryWithPrefixToAdd,
 )
-from pants.engine.isolated_process import ExecuteProcessResult, MultiPlatformExecuteProcessRequest
+from pants.engine.isolated_process import (
+  ExecuteProcessRequest,
+  ExecuteProcessResult,
+  MultiPlatformExecuteProcessRequest,
+)
+from pants.engine.legacy.structs import PythonTargetAdaptor, TargetAdaptor
 from pants.engine.platform import Platform, PlatformConstraint
-from pants.engine.rules import optionable_rule, rule
+from pants.engine.rules import RootRule, optionable_rule, rule
 from pants.engine.selectors import Get
+from pants.util.strutil import create_path_env_var
+
+
+@dataclass(frozen=True)
+class PexRequirements:
+  requirements: Tuple[str] = ()
+
+  @classmethod
+  def create_from_adaptors(cls, adaptors: Tuple[TargetAdaptor], additional_requirements: Tuple[str]) -> 'PexRequirements':
```

Comment:
This needs to be `Tuple[TargetAdaptor, ...]` and `Tuple[str, ...]`. Otherwise, you are saying a Tuple with only one single element. The `...` signifies the length can be arbitrarily long.

Reply:
Right, that non-intuitive thing about Python tuple type annotation, will fix.